src/server/services/pipeline-in-progress/handleCumulative.js

```diff
@@ -0,0 +1,52 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+const moment = require('moment');
+
+/**
+ * Handle cumulative data, assume that last row is the first reading (skip this row).
+ * @Example
+ * 	row 0: reading #0 + #1 + #2
+ *  row 1: reading #0 + #1
+ * 	row 2: reading #0
+ * => reading #1 = row 0 - row 1
+ *    reading #2 = row 1 - row 2
+ *    reading #0 is cummulative value from unknown readings that are may or may not inserted before
+ * @param {2d array} rows
+ * @param readingRepetition value is 1 if reading is not duplicated. 2 if repeated twice and so on (E-mon D-mon meters)
+ */
+
+function handleCumulative(rows, readingRepetition) {
+	const result = [];
+	// Initialize timestamps and other variables
+	let startTimestamp = moment(0);
+	let endTimestamp = moment(0);
+	let meterReading = 0;
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
+	for (let index = 0; index < rows.length; ++index) {
+		// To read data where same reading is repeated. Like E-mon D-mon meters
+		if (index !== 0 && Math.abs((index - readingRepetition) % readingRepetition) === 0) {
+			// set start_timestamp and end_timestamp
+			startTimestamp = moment(rows[index][1], 'MM/DD/YY HH:mm');
+			endTimestamp = moment(rows[index - readingRepetition][2], 'MM/DD/YY HH:mm');
+			// meterReading
+			meterReading1 = rows[index - readingRepetition][0];
+			meterReading2 = rows[index][0];
+			meterReading = meterReading1 - meterReading2;
+			// To handle cumulative readings that resets at midnight
+			if (meterReading < 0) {
```

Comment:
This is a really good point. I'm just not sure I fully understand your solution. Can you please clarify what you mean by "check if the reading(s) are negative and then not do this process"? Are you referring to the actual reading values (meterReading1, meterReading2)? I may be missing something since I don't really see how this would solve the problem (meterReading1 and meterReading2 can both be non-negative/valid to generate a negative/invalid difference).

I do like the idea of having an "allow reset" parameter though, I think we can also let the user define the time period to accept "reset" reading value (e.g. within 1 hour after midnight). If we have that information, we can simply check for the timestamp every time we get a negative value and immediately accept or reject it accordingly. My only concern is whether we're asking too much from the user, but I think having a nice "default" setting may help

Reply:
I was referring to meterReadings1 or 2. My thinking was that if either is negative then we don't do the wrap (the if that resets the value) since that can take a negative reading value and reset it to positive. The only negative reading I know of is for temperature and that is not done cumulatively so it should not be in this code. Thus, something should be wrong with a negative value so trying to take differences is problematic. The concern is that a negative reading might be covered up since the reset could make it positive. I agree with you that two positive values can lead to a negative difference but that is actually an error in cumulative data so it should be left and detected later. Given this, I propose that the check be changed to
if (meterReading1 < 0 || meterReading < 0) {
This means a negative reading is passed along and not reset. I don't think we need to check 
meterReading2 (assuming my limited analysis is right) since it will just make it more positive (the difference) and it would be detected on the next loop cycle when it becomes meterReading1.

I'm not sure the perfect default value for how close to midnight exists. I think using one hour seems fine. This means we are much less likely to reset bad values since they will not be near midnight. I agree this asks something of the user but the default means it will likely work without them doing anything. I think this would be a good change.

Reply:
I think I got your idea now, thank you for the clarification! Just to confirm, I'm understanding that there are 4 different cases that need to be handled differently in handleCumulative:

1. Positive difference between 2 positive readings: accept
2. Positive difference between 2 negative readings: reject immediately
3. Negative difference between 2 positive readings, happens at midnight: convert difference to meterReading1, accept as a reset
4. Negative difference between 2 positive readings, happens anytime else: reject eventually in validateReadings

For case 2, my solution is to reject the whole set of readings (since an invalid cumulative value would shift all the others, accepting partially may be messy). The condition is meterReading1 < 0

For case 3, I think we can check for the date difference between the start and endTimeStamp to see if the value has been read overnight. This way, there's no need for user input nor a default "midnight threshold." The condition is meterReading < 0 && endTimestamp.isAfter(startTimestamp, 'date')

Am I missing anything? I've added code to implement these new catches and a little addition to handleCumulativeTests, please let me know if you notice any problem.

Thank you!